### DISCUSSION TITLE
[@types/snarkjs] Wrong return type on `r1cs.info`

### DISCUSSION THREAD

**Discussion by erhant** [DefinitelyTyped/DefinitelyTyped#68203] [Issues with a @types package] [answered]
Greetings @cedoor ! (EDIT: I had tagged you manually but apparently TS bot does that too)

I believe the type `R1CSInfoType` has been borrowed from [Circomkit](https://github.com/erhant/circomkit), as per this [comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66703#pullrequestreview-1627412688). However, `R1CSInfoType` was a wrapper around the actual underlying type returned by snarkjs, but with slightly better variable naming. 

Here is the related function under [`r1cs_info.js`](https://github.com/iden3/snarkjs/blob/master/src/r1cs_info.js#L26), which calls some more functions in the background. When we do a loop over `Object.keys` for the returned value and print the key & its `typeof` we see the following:

```sh
n8 number
prime bigint
curve object
F object
nVars number
nOutputs number
nPubInputs number
nPrvInputs number
nLabels number
nConstraints number
useCustomGates boolean
constraints object
customGates object
customGatesUses object
```

In the end, we should _at least_ be returning the following type:

```ts
type R1CSInfoType = {
  n8: number;
  nVars: number;
  nConstraints: number;
  nPrvInputs: number;
  nPubInputs: number;
  useCustomGates: boolean;
  nLabels: number;
  nOutputs: number;
  prime: bigint;
};
```

The following fields are more complicated & I don't know their types in full yet:

- `F` is a `WasmField1` instance
- `curve` is an object with quite a lot of properties, such as `WasmField1` instances for its fields, and some properties about the curve (`q`, `r` and its name)
- Not sure about custom gates yet.
- `constraints` is an array (of length `nConstraints`) with triples (i.e. A, B, C) that are in form:
```ts
Record<`${number}`, Uint8Array> 
```

Perhaps the fields above could be returned with type `object` or `any` in the worst case.

> [!TIP]
>
> <sub><i>oh god I wish we re-wrote SnarkJS with types & using Bun runtime<sub>would love to do that together hahah</i></sub></sub>

[a comment by an automated account is not shown]

**Comment by cedoor** [comment]
Hey @erhant, thanks for pointing this out!

Yes, I copied that type without double-checking 😄 

I think it should include all the parameters. For the more complex ones we could use `any` and add a link to the JS object. Like here:  https://github.com/DefinitelyTyped/DefinitelyTyped/blob/68a06dbd340bd1a6442f1d22d01ac071e7ec9dd1/types/circomlibjs/index.d.cts#L16

What do you think?

It would be great to have types in SnarkJS by default, and especially modular and documented code. Not at all trivial to re-implement it though 😢

**Comment by erhant** [comment]
> I think it should include all the parameters. For the more complex ones we could use any and add a link to the JS object.

I agree with this, should leave that much detail for later! Perhaps those underlying libraries (such as ffjavascript) can have type declarations at some point too, which would solve the problem here.

> It would be great to have types in SnarkJS by default, and especially modular and documented code. Not at all trivial to re-implement it though 😢

Indeed... My half-joke comment aside, I would really want to see Bun compatibility with SnarkJS at some point, in such a computation-heavy task that is proof generation, Bun could do wonders for us; for testing as well! I checked this sometime ago, there is a problem (https://github.com/iden3/fastfile/issues/91) but it definitely seems doable. Food for thought :)

**Comment by cedoor** [comment]
> I agree with this, should leave that much detail for later! Perhaps those underlying libraries (such as ffjavascript) can have type declarations at some point too, which would solve the problem here.

Agree! 

> Indeed... My half-joke comment aside, I would really want to see Bun compatibility with SnarkJS at some point, in such a computation-heavy task that is proof generation, Bun could do wonders for us; for testing as well! I checked this sometime ago, there is a problem ([iden3/fastfile#91](https://github.com/iden3/fastfile/issues/91)) but it definitely seems doable. Food for thought :)

We've just checked it and we found the same issue 😄 I guess we still need to wait.. or solve it ourselves eventually. But yeah it would be nice to use Bun!

**Comment by cedoor** [comment, marked as answer]
Re: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68206
